url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dane/DaneVerifier.java

```diff
@@ -0,0 +1,273 @@
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A helper class to validate the usage of TLSA records.
+ */
+public class DaneVerifier {
+    private final static Logger LOGGER = Logger.getLogger(DaneVerifier.class.getName());
+
+    private final AbstractDNSClient client;
+
+    public DaneVerifier() {
+        this(new DNSSECClient());
+    }
+
+    public DaneVerifier(AbstractDNSClient client) {
+        this.client = client;
+    }
+
+    /**
+     * Verifies the certificate chain in an active {@link SSLSocket}. The socket must be connected.
+     *
+     * @param socket A connected {@link SSLSocket} whose certificate chain shall be verified using DANE.
+     * @return Whether the DANE verification is the only requirement according to the TLSA record.
+     * If this method returns {@code false}, additional PKIX validation is required.
+     * @throws CertificateException if the certificate chain provided differs from the one enforced using DANE.
+     */
+    public boolean verify(SSLSocket socket) throws CertificateException {
+        if (!socket.isConnected()) {
+            throw new IllegalStateException("Socket not yet connected.");
+        }
+        return verify(socket.getSession());
+    }
+
+    /**
+     * Verifies the certificate chain in an active {@link SSLSession}.
+     *
+     * @param session An active {@link SSLSession} whose certificate chain shall be verified using DANE.
+     * @return Whether the DANE verification is the only requirement according to the TLSA record.
+     * If this method returns {@code false}, additional PKIX validation is required.
+     * @throws CertificateException if the certificate chain provided differs from the one enforced using DANE.
+     */
+    public boolean verify(SSLSession session) throws CertificateException {
+        try {
+            return verifyCertificateChain(convert(session.getPeerCertificateChain()), session.getPeerHost(), session.getPeerPort());
+        } catch (SSLPeerUnverifiedException e) {
+            throw new CertificateException("Peer not verified", e);
+        }
+    }
+
+    /**
+     * Verifies a certificate chain to be valid when used with the given connection details using DANE.
+     *
+     * @param chain A certificate chain that should be verified using DANE.
+     * @param hostName The DNS name of the host this certificate chain belongs to.
+     * @param port The port number that was used to reach the server providing the certificate chain in question.
+     * @return Whether the DANE verification is the only requirement according to the TLSA record.
+     * If this method returns {@code false}, additional PKIX validation is required.
+     * @throws CertificateException if the certificate chain provided differs from the one enforced using DANE.
+     */
+    public boolean verifyCertificateChain(X509Certificate[] chain, String hostName, int port) throws CertificateException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 112-122 into a separate method.